content/fundamentals/functions.py

```diff
@@ -1,179 +1,330 @@
-# - The `zip` method allows iterating over multiple lists at the same time.
-#   If the inputs have different lengths, the iteration will stop once it reaches the end of the shortest list.
-#
-# - We use `typing` (e.g. `conditions: list[bool]`) in the signature of the function to
-#   specify the type required for the input arguments. The type is not enforced on runtime,
-#   but mainly good-practice as it can be used for code analysis and documentation.
+    Extract elements from a list where the corresponding condition is True.
 
+    :param values: List of values to filter
+    :param conditions: List of True/False values (must be same length as values)
+    :return: A list containing only the values where the condition was True
+    """
+    result = []
+
+    # Loop through both values and conditions simultaneously
+    for value, condition in zip(values, conditions): # `zip` pairs elements from both lists
+        if condition:  # Check if the condition is True
+            result.append(value)  # Add the value to the result if condition is True
+
+    return result
+#%% md
+# The `if` statement checks a condition and only executes the indented code
+# below it if the condition is `True`.
+# 
+# The `zip()` function is very useful - it allows us to iterate through
+# multiple lists at the same time, pairing elements together.
+# 
+# Now let's use this function to get the depths where our high-grade samples were found:
+#%%
+# Get depths where grade is above 1.0
+high_grade_depths = get_matching_values(depths, high_grade_indicators)
+print("Depths with high-grade samples:", high_grade_depths)
+
+#%% md
+# ## Function Type Hints
+# 
+# Python allows you to add **type hints** to your functions. Type hints are not
+#  enforced by Python, but they help document the expected types of parameters
+#  and return values. This makes your code easier to understand and helps code
+#  analysis tools catch potential errors:
+#%%
+def get_matching_values(values: list, conditions: list[bool]) -> list:
+    """
+    Extract elements from a list where the corresponding condition is True.
 
-# Calling the `get_depths` method with the `grades` and `logic` previously computed we get
+    :param values: List of values to filter
+    :param conditions: List of True/False values (must be same length as values)
+    :return: A list containing only the values where the condition was True
+    """
+    result = []
+
+    for value, condition in zip(values, conditions):
+        if condition:
+            result.append(value)
+
+    return result
+#%% md
+# Type hints are especially helpful in larger projects where you want to make
+# sure your functions are used correctly.
+#%% md
+# ## Putting It All Together
+# Now we can combine our functions to solve our original problem - finding the depths where there are high-grade samples:
+#%%
+def find_high_grade_locations(grades, depths, threshold=1.0):
```

Comment:
```suggestion
def find_high_grade_locations(grades: list, depths: list, threshold: float = 1.0) -> list:
```
Since we've introduced Typing above, we should probably demo it here.